Fixing requirePaddingNewLinesAfterBlocks to no longer fail on argument lists

TheSavior opened the pull request:
Fixes #1232

zxqfox commented on lib/rules/require-padding-newlines-after-blocks.js:
````diff
@@ -1,14 +1,20 @@
  *
  * ```js
  * "requirePaddingNewLinesAfterBlocks": true
+ * "requirePaddingNewLinesAfterBlocks": { "allExcept": ["inCallExpressions", "inArrayExpressions"]}
````
Space before closing brace?

TheSavior replied:
Hah. Good catch. Fixed.

markelog commented:
`inCallExpressions` and `inArrayExpressions` are specific options for this rule only, i wouldn't want to invent any new names if there wouldn't be any other way to do this.

#1232 could be solved much more easily

TheSavior commented:
@markelog, if your PR passes the added tests in this PR I'm okay with your solution. From an initial glance, I am guessing that it won't.

TheSavior commented:
Also referencing #1237 here.

markelog commented:
See the latest changes/comments in https://github.com/jscs-dev/node-jscs/pull/1246

TheSavior commented:
It is the same issue with arrays though

```javascript
var a = [
2,
3,
function() {
} // this will expect a blank line after it.
]
```

TheSavior commented:
I guess I'm not really sure what the issue with the fix in this PR is. Is it that you don't want any exclude options at all, only `true` or `false`? Or is it that you think we should change the defaults?

markelog commented:
@TheSavior i'm talking about consistency - `allExcept` values, in other rules, either a token list or list of exceptions, here we trying to provide rule options through it. Inventing our own node types or token are not ideal.

Plus why would we need to complicate the rule, if we could do it much more simpler?

markelog commented:
Sorry @TheSavior, it sometimes awful not to land a nice piece of work :-(. I'm closing this in favour of #1246

That being said, i think we can add support for arrays though.

TheSavior commented:
I'm not so much concerned about having to throw code away, but mostly the support of arrays. I was the original author of the rule, and discovered when running this that it was breaking in a bunch of places in our codebase that it shouldn't have been. I repeatedly stated on both of our PRs, that this rule needed to pass the tests I added and for some reason arrays were punted as not important enough for the hotfix. 

Frankly, I still can't use this rule until the array fix is in there, and that is what makes me annoyed. Especially as the original author of the rule.

(Obviously I can use a custom rule in our project for the time being, but that is awfully disappointing).

mikesherov commented:
@markelog we should land support for arrays too.

markelog commented:
I think we should go further to be consistent there and ignore blocks in objects too, if there is need for that functionality we would need a different rule altogether.

mikesherov commented:
Agreed.

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -72,10 +113,30 @@ module.exports = function() {};
+        this._allowInCallExpressions = false;
+        this._allowInArrayExpressions = false;
+
+        if (typeof value === 'object') {
```
This whole block is very complicated, i guess it would be better to work on https://github.com/jscs-dev/node-jscs/issues/895 instead?

TheSavior commented:
Alright, I added the `isException` helper into this branch and added an option for object properties as mentioned in #1260. 

@markelog would you like to reopen this PR so it grabs the new commit, or should I create a new PR?

markelog commented:
Did you push these changes?

markelog commented:
I see falling tests in CI, could you check it out?

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -109,11 +197,11 @@ module.exports.prototype = {
                 return;
             }
-        });
+        }).bind(this));
```
We usually create a variable for this

markelog commented:
How about opposite rule?

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -32,9 +39,54 @@
+ * var foo = {
+ *     a: 2,
+ *     b: function() {
+ *     }
```
comma

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
````diff
@@ -1,14 +1,21 @@
  *
  * ```js
  * "requirePaddingNewLinesAfterBlocks": true
+ * "requirePaddingNewLinesAfterBlocks": { "allExcept": ["inCallExpressions", "inArrayExpressions", "inProperties"] }
````
Couple new lines would help here

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
````diff
@@ -32,9 +39,54 @@
+ *     }
+ * ];
+ * ```
+* ##### Valid for `{ "allExcept": "inProperties" }`
````
shouldn't this be an array?

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -32,9 +39,54 @@
+ *
+ *     );
+ *
+ *     var a = [
```
Maybe more meaningful examples would fit better

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -67,15 +111,55 @@ var excludes = {
+
+            value.allExcept.forEach((function(except) {
+                if (except === 'inCallExpressions') {
+                    exceptions.CallExpression = true;
```
Nice solution

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -67,15 +111,55 @@ var excludes = {
-            this.getOptionName() + ' option requires a true value or should be removed'
-        );
+        if (typeof value === 'object') {
+            assert(Array.isArray(value.allExcept), this.getOptionName() + ' option requires "allExcept" ' +
```
You can call `getOptionName` once assigning it to variable

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -67,15 +111,55 @@ var excludes = {
+                    exceptions.Property = true;
+                } else {
+                    assert(false, this.getOptionName() + ' option requires "allExcept" to only have ' +
+                        '`inCallExpressions` or `inArrayExpressions`');
```
Seriously, why does everybody use " ` " sign?

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -67,15 +111,55 @@ var excludes = {
+                    assert(false, this.getOptionName() + ' option requires "allExcept" to only have ' +
+                        '`inCallExpressions` or `inArrayExpressions`');
+                }
+            }).bind(this));
```
you can pass context as second argument to the `forEach`

markelog commented on lib/rules/require-padding-newlines-after-blocks.js:
```diff
@@ -67,15 +111,55 @@ var excludes = {
+            assert(value.allExcept.length > 0, this.getOptionName() + ' option requires "allExcept" ' +
+                'to have at least one item or be set to `true`');
+
+            value.allExcept.forEach((function(except) {
```
`((` really subjective

TheSavior commented:
Sorry for taking so long to get to updating this PR. @markelog I fixed the things you mentioned.

I need to look into why this build is failing more. When I run `npm test` on my computer I don't get any errors.

markelog commented:
Code is simplified, failures are fixed
Thank you.
